### PR TITLE
Support custom errors that pass in lower-level errors.

### PR DESCRIPTION
**Pull request by mrutkows**
Fixes https://github.com/apache/incubator-openwhisk-wskdeploy/issues/648

### CODE REVIEW

**Comment by pritidesai** [comment]
Very cool :)

```
 ./wskdeploy -m tests/src/integration/managed-deployment/my-manifest.yaml --managed
Error: servicedeployer.go [136]: [ERROR_YAML_FILE_FORMAT_ERROR]: File: [manifest.yaml]: Project name in manifest file is mandatory for managed deployments
```

**Comment by mrutkows** [comment]
Please see issue https://github.com/apache/incubator-openwhisk-wskdeploy/issues/650 wheich describe refactoring needed to allow better YAML parser unit tests.

**Comment by mrutkows** [comment]
Please see issue https://github.com/apache/incubator-openwhisk-wskdeploy/issues/650 which describe refactoring needed to allow better YAML parser unit tests.

File: deployers/manifestreader.go [hunk 1]
```diff
@@ -46,15 +46,17 @@ func (deployer *ManifestReader) ParseManifest() (*parsers.YAML, *parsers.YAMLPar
 	manifest, err := manifestParser.ParseManifest(dep.ManifestPath)
 
 	if err != nil {
-		return manifest, manifestParser, utils.NewFileReadError(dep.ManifestPath, err.Error())
+		//return manifest, manifestParser, utils.NewFileReadError(dep.ManifestPath, err.Error())
```
Comment (houshengbo) [inline deployers/manifestreader.go]: Shall we remove this line?

File: deployers/manifestreader.go [hunk 2]
```diff
@@ -46,15 +46,17 @@ func (deployer *ManifestReader) ParseManifest() (*parsers.YAML, *parsers.YAMLPar
...
+		return manifest, manifestParser, err
 	}
 	return manifest, manifestParser, nil
 }
 
 func (reader *ManifestReader) InitRootPackage(manifestParser *parsers.YAMLParser, manifest *parsers.YAML, ma whisk.KeyValue) error {
 	packages, err := manifestParser.ComposeAllPackages(manifest, reader.serviceDeployer.ManifestPath, ma)
 	if err != nil {
-		return utils.NewYAMLFileFormatError(manifest.Filepath, err.Error())
+		//return utils.NewYAMLFileFormatError(manifest.Filepath, err.Error())
```
Comment (houshengbo) [inline deployers/manifestreader.go]: Shall we remove this line?

File: utils/wskdeployerror_test.go [hunk 3]
```diff
@@ -133,52 +137,62 @@ func TestCustomErrorOutputFormat(t *testing.T) {
...
-		"\n==> Line [123]: found duplicate %YAML directive"
-
-	expectedResult = fmt.Sprintf("%s [%d]: [%s]: " + FILE + ": [%s]: %s",
-		packageName,
-		err10.LineNum,
-		ERROR_YAML_PARSER_ERROR,
-		filepath.Base(TEST_EXISTANT_MANIFEST_FILE),
-		msgs)
-	assert.Equal(t, expectedResult, actualResult, "Expected [" + expectedResult + "] but got [" + actualResult + "]")
+	// TODO add a unit test once we re-factor error related modules into a new package
```
Comment (houshengbo) [inline utils/wskdeployerror_test.go]: Why do you comment out this lines? Is there going to be new test case replacing the old one?

Comment (mrutkows) [reply utils/wskdeployerror_test.go]: This is intentional see referenced issue 650.  I do not want to refactor using this issue.

**Comment by houshengbo** [comment]
@mrutkows Two tests failed in Travis due to "Concurrent modification to resource detected". You probably need to change the entity name in the yaml files for the tests.

**Comment by mrutkows** [comment]
@houshengbo these errors are not mine... these are concurrency issues (that are happening just today).  AND they are moving around between different testcases.  As you saw, the PR built fine yesterday; today I just removed 2 comment lines. NO CODE CHANGES

**Comment by mrutkows** [comment]
opened issue https://github.com/apache/incubator-openwhisk-wskdeploy/issues/651
which showed has different errors on different tests after each restart of Travis.

and after a 3rd restart... the errors moved again...
```
=== RUN   TestInvalidActionAnnotations

==> Executing: wskdeploy -m /home/travis/gopath/src/github.com/apache/incubator-openwhisk-wskdeploy/tests/src/integration/validate-action-annotations/manifest.yaml -d /home/travis/gopath/src/github.com/apache/incubator-openwhisk-wskdeploy/tests/src/integration/validate-action-annotations/deployment-with-invalid-annotations.yaml

==> Output: WARNING: The authentication key was retrieved from whisk.properties which will soon be deprecated please do not use it outside of Travis builds.

WARNING: The API host was retrieved from whisk.properties which will soon be deprecated please do not use it outside of Travis builds.

.

==> Error: wskdeploy.go [125]: [ERROR_COMMAND_FAILED]: Command: [wskdeploy]: Error: deploymentreader.go [241]: [ERROR_YAML_FILE_FORMAT_ERROR]: File: [deployment-with-invalid-annotations.yaml]: 

==> Annotation key "action_annotation_5" does not exist in manifest file but specified in deployment file.

```

**Comment by mrutkows** [comment]
Need to look at the failing tests and grep to see any of their package/action/trigger/rule names are duplicated in other places as a first step.

**Comment by houshengbo** [comment]
@mrutkows I believe that the recent changes to openwhisk core have led to the concurrent issues, since I am using the latest openwhisk to deploy, wskdeploy will randomly raise errors for different cases.

Let's find out the last good commit for openwhisk core.

**Pull request merged**